<h2 align='center' style='color:blue'>Dropout Regularization In Deep Neural Network</h2>

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [ ]:
import warnings
warnings.filterwarnings('ignore')

In [ ]:
df = pd.read_csv("/content/sample_data/sonar.all-data.csv", header=None)
df.sample(5)

0       1       2       3       4       5       6       7       8   \
9    0.0164  0.0173  0.0347  0.0070  0.0187  0.0671  0.1056  0.0697  0.0962   
113  0.0283  0.0599  0.0656  0.0229  0.0839  0.1673  0.1154  0.1098  0.1370   
21   0.0664  0.0575  0.0842  0.0372  0.0458  0.0771  0.0771  0.1130  0.2353   
62   0.0086  0.0215  0.0242  0.0445  0.0667  0.0771  0.0499  0.0906  0.1229   
80   0.0229  0.0369  0.0040  0.0375  0.0455  0.1452  0.2211  0.1188  0.0750   

         9   ...      51      52      53      54      55      56      57  \
9    0.0251  ...  0.0090  0.0223  0.0179  0.0084  0.0068  0.0032  0.0035   
113  0.1767  ...  0.0147  0.0170  0.0158  0.0046  0.0073  0.0054  0.0033   
21   0.1838  ...  0.0141  0.0190  0.0043  0.0036  0.0026  0.0024  0.0162   
62   0.1185  ...  0.0072  0.0054  0.0022  0.0016  0.0029  0.0058  0.0050   
80   0.1631  ...  0.0064  0.0022  0.0122  0.0151  0.0056  0.0026  0.0029   

         58      59  60  
9    0.0056  0.0040   R  
113  0.0045  0.0079   M  
21   0.0109  0.0079   R  
62   0.0024  0.0030   R  
80   0.0104  0.0163   R  

[5 rows x 61 columns]

In [ ]:
df.shape

(208, 61)

In [ ]:
# check for nan values
df.isna().sum()

0     0
1     0
2     0
3     0
4     0
     ..
56    0
57    0
58    0
59    0
60    0
Length: 61, dtype: int64

In [ ]:
df.columns

Int64Index([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
            17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
            34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50,
            51, 52, 53, 54, 55, 56, 57, 58, 59, 60],
           dtype='int64')

In [ ]:
df[60].value_counts() # label is not skewed

M    111
R     97
Name: 60, dtype: int64

In [ ]:
X = df.drop(60, axis=1)
y = df[60]
y.head()

0    R
1    R
2    R
3    R
4    R
Name: 60, dtype: object

In [ ]:
y = pd.get_dummies(y, drop_first=True)
y.sample(5) # R --> 1 and M --> 0

R
47   1
175  0
169  0
55   1
163  0

In [ ]:
y.value_counts()

R
0    111
1     97
dtype: int64

In [ ]:
X.head()

0       1       2       3       4       5       6       7       8   \
0  0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109   
1  0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
2  0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   
3  0.0100  0.0171  0.0623  0.0205  0.0205  0.0368  0.1098  0.1276  0.0598   
4  0.0762  0.0666  0.0481  0.0394  0.0590  0.0649  0.1209  0.2467  0.3564   

       9   ...      50      51      52      53      54      55      56  \
0  0.2111  ...  0.0232  0.0027  0.0065  0.0159  0.0072  0.0167  0.0180   
1  0.2872  ...  0.0125  0.0084  0.0089  0.0048  0.0094  0.0191  0.0140   
2  0.6194  ...  0.0033  0.0232  0.0166  0.0095  0.0180  0.0244  0.0316   
3  0.1264  ...  0.0241  0.0121  0.0036  0.0150  0.0085  0.0073  0.0050   
4  0.4459  ...  0.0156  0.0031  0.0054  0.0105  0.0110  0.0015  0.0072   

       57      58      59  
0  0.0084  0.0090  0.0032  
1  0.0049  0.0052  0.0044  
2  0.0164  0.0095  0.0078  
3  0.0044  0.0040  0.0117  
4  0.0048  0.0107  0.0094  

[5 rows x 60 columns]

In [ ]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=1)

In [ ]:
X_train.head()

0       1       2       3       4       5       6       7       8   \
67   0.0368  0.0403  0.0317  0.0293  0.0820  0.1342  0.1161  0.0663  0.0155   
14   0.0124  0.0433  0.0604  0.0449  0.0597  0.0355  0.0531  0.0343  0.1052   
164  0.0163  0.0198  0.0202  0.0386  0.0752  0.1444  0.1487  0.1484  0.2442   
179  0.0394  0.0420  0.0446  0.0551  0.0597  0.1416  0.0956  0.0802  0.1618   
19   0.0126  0.0149  0.0641  0.1732  0.2565  0.2559  0.2947  0.4110  0.4983   

         9   ...      50      51      52      53      54      55      56  \
67   0.0506  ...  0.0058  0.0091  0.0160  0.0160  0.0081  0.0070  0.0135   
14   0.2120  ...  0.0078  0.0083  0.0057  0.0174  0.0188  0.0054  0.0114   
164  0.2822  ...  0.0027  0.0077  0.0026  0.0031  0.0083  0.0020  0.0084   
179  0.2558  ...  0.0118  0.0146  0.0040  0.0114  0.0032  0.0062  0.0101   
19   0.5920  ...  0.0153  0.0092  0.0035  0.0098  0.0121  0.0006  0.0181   

         57      58      59  
67   0.0067  0.0078  0.0068  
14   0.0196  0.0147  0.0062  
164  0.0108  0.0083  0.0033  
179  0.0068  0.0053  0.0087  
19   0.0094  0.0116  0.0063  

[5 rows x 60 columns]

# Using Deep Learning Model

## Model without Dropout Layer

In [ ]:
import tensorflow as tf
from tensorflow import keras

In [ ]:
model = keras.Sequential([
    keras.layers.Dense(60, input_dim=60, activation='relu'),
    keras.layers.Dense(30, activation='relu'),
    keras.layers.Dense(15, activation='relu'),
    keras.layers.Dense(1, activation='sigmoid')
])

model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

model.fit(X_train, y_train, epochs=100, batch_size=8)

Epoch 1/100
20/20 [==============================] - 1s 2ms/step - loss: 0.6998 - accuracy: 0.5064
Epoch 2/100
20/20 [==============================] - 0s 3ms/step - loss: 0.6824 - accuracy: 0.6154
Epoch 3/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6716 - accuracy: 0.6859
Epoch 4/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6569 - accuracy: 0.6731
Epoch 5/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6356 - accuracy: 0.6859
Epoch 6/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6226 - accuracy: 0.6731
Epoch 7/100
20/20 [==============================] - 0s 2ms/step - loss: 0.5992 - accuracy: 0.6731
Epoch 8/100
20/20 [==============================] - 0s 2ms/step - loss: 0.5600 - accuracy: 0.7628
Epoch 9/100
20/20 [==============================] - 0s 2ms/step - loss: 0.5560 - accuracy: 0.7179
Epoch 10/100
20/20 [==============================] - 0s 2ms/step - loss: 0.4913 - accuracy: 0.8013
Epoch 11/

In [ ]:
model.evaluate(X_test, y_test)

2/2 [==============================] - 0s 10ms/step - loss: 1.0097 - accuracy: 0.7500


[1.0096544027328491, 0.75]

Training Accuracy >>> Test Accuracy

In [ ]:
y_pred = model.predict(X_test).reshape(-1)
print(y_pred[:10])

# round the values to nearest integer ie 0 or 1
y_pred = np.round(y_pred)
print(y_pred[:10])

[2.7657171e-07 9.6628594e-01 9.6729136e-01 1.4380545e-05 9.9998510e-01
 9.9837756e-01 1.9437099e-01 9.9999714e-01 2.9569921e-05 9.9999380e-01]
[0. 1. 1. 0. 1. 1. 0. 1. 0. 1.]


In [ ]:
y_test[:10]

R
186  0
155  0
165  0
200  0
58   1
34   1
151  0
18   1
202  0
62   1

In [ ]:
from sklearn.metrics import confusion_matrix , classification_report

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.72      0.85      0.78        27
           1       0.80      0.64      0.71        25

    accuracy                           0.75        52
   macro avg       0.76      0.75      0.75        52
weighted avg       0.76      0.75      0.75        52



### Model with Dropout Layer

In [ ]:
modeld = keras.Sequential([
    keras.layers.Dense(60, input_dim=60, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(30, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(15, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(1, activation='sigmoid')
])

modeld.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

modeld.fit(X_train, y_train, epochs=100, batch_size=8)

Epoch 1/100
20/20 [==============================] - 1s 3ms/step - loss: 0.7063 - accuracy: 0.5385
Epoch 2/100
20/20 [==============================] - 0s 3ms/step - loss: 0.7072 - accuracy: 0.5064
Epoch 3/100
20/20 [==============================] - 0s 3ms/step - loss: 0.6963 - accuracy: 0.5449
Epoch 4/100
20/20 [==============================] - 0s 3ms/step - loss: 0.6872 - accuracy: 0.5256
Epoch 5/100
20/20 [==============================] - 0s 3ms/step - loss: 0.7174 - accuracy: 0.4744
Epoch 6/100
20/20 [==============================] - 0s 3ms/step - loss: 0.6926 - accuracy: 0.4808
Epoch 7/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6937 - accuracy: 0.5705
Epoch 8/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6826 - accuracy: 0.5577
Epoch 9/100
20/20 [==============================] - 0s 3ms/step - loss: 0.6857 - accuracy: 0.5256
Epoch 10/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6767 - accuracy: 0.6410
Epoch 11/

In [ ]:
modeld.evaluate(X_test, y_test)

2/2 [==============================] - 0s 6ms/step - loss: 0.4283 - accuracy: 0.7885


[0.42828860878944397, 0.7884615659713745]

Training Accuracy is still good but Test Accuracy Improved

In [ ]:
y_pred = modeld.predict(X_test).reshape(-1)
print(y_pred[:10])

# round the values to nearest integer ie 0 or 1
y_pred = np.round(y_pred)
print(y_pred[:10])

[7.4116142e-05 5.5097592e-01 8.4000230e-01 6.4998567e-03 9.9800932e-01
 8.7181175e-01 3.1416076e-01 9.9771810e-01 1.0738313e-02 9.9874419e-01]
[0. 1. 1. 0. 1. 1. 0. 1. 0. 1.]


In [ ]:
from sklearn.metrics import confusion_matrix , classification_report

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.74      0.93      0.82        27
           1       0.89      0.64      0.74        25

    accuracy                           0.79        52
   macro avg       0.81      0.78      0.78        52
weighted avg       0.81      0.79      0.78        52



**You can see that by using dropout layer test accuracy increased from 0.77 to 0.81**